app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorEventRepo.kt

```diff
@@ -0,0 +1,113 @@
+         * content in that direction.
+         */
+        data class ScrolledToEdge(val edge: Direction) : CursorEvent()
+        data class CursorMoved(val direction: Direction) : CursorEvent()
+    }
+
+    private var cursorController: CursorController? = null
+    private val keyEvents: Subject<KeyEvent> = PublishSubject.create()
+
+    fun setCursorController(cursorController: CursorController) {
+        this.cursorController = cursorController
+    }
+
+    fun pushKeyEvent(keyEvent: KeyEvent) {
+        keyEvents.onNext(keyEvent)
+    }
+
+    val webRenderDirectionEvents: Observable<CursorEvent> by lazy {
+        fun <T> Observable<T>.throttleEvents(): Observable<T> =
+                this.throttleFirst(10, TimeUnit.MILLISECONDS)
+
+        fun <T> Observable<Pair<T, ScreenControllerStateMachine.ActiveScreen>>.onlyEmitIfWebRenderIsActive() =
+                this.filter { (_, activeScreen) -> activeScreen == ScreenControllerStateMachine.ActiveScreen.WEB_RENDER }
+
+        fun <T> Observable<Pair<T, ScreenControllerStateMachine.ActiveScreen>>.dropActiveScreen() =
+                this.map { (value, _) -> value }
+
+        fun Observable<KeyEvent>.filterIsDirectionKey() =
+                this.filter { DIRECTION_KEYS.contains(it.keyCode) }
+
+        fun Observable<KeyEvent>.filterIsKeyDown() =
+                // i.e., key is pressed, not released
+                this.filter { it.action == ACTION_DOWN }
+
+        fun Observable<KeyEvent>.mapEventToDirection() =
+                this.flatMap {
+                    when (it.keyCode) {
+                        KEYCODE_DPAD_UP -> Observable.just(Direction.UP)
+                        KEYCODE_DPAD_DOWN -> Observable.just(Direction.DOWN)
+                        KEYCODE_DPAD_LEFT -> Observable.just(Direction.LEFT)
+                        KEYCODE_DPAD_RIGHT -> Observable.just(Direction.RIGHT)
+                        else -> Observable.empty()
+                    }
+                }
+
+        fun Observable<Direction>.mapToCursorEvent() =
+                this.map {
+                    val scrollToEdgeEvent = when {
+                        it == Direction.UP &&
+                                cursorController?.cursorIsNearTopOfScreen() == true &&
+                                cursorController?.webViewCouldScrollUp() == false
+                        -> CursorEvent.ScrolledToEdge(Direction.UP)
+
+                        it == Direction.DOWN &&
+                                cursorController?.cursorIsNearBottomOfScreen() == true &&
+                                cursorController?.webViewCouldScrollDown() == false
+                        -> CursorEvent.ScrolledToEdge(Direction.DOWN)
+
+                        else -> null
+                    }
+                    return@map scrollToEdgeEvent ?: CursorEvent.CursorMoved(it)
+                }
+
+        Observables.combineLatest(keyEvents, screenController.currentActiveScreen)
+                .throttleEvents()
```

Comment:
+1000000

Reply:
I did just notice that there's a small problem here, though.  Since we're throttling after we combine, we could potentially drop screen change events, and we probably want to respond to all of them.  I've updated the code to `Observables.combineLatest(keyEvents.throttleEvents(), screenController.currentActiveScreen)`